src/python/Services/ReqMgr/ReqMgrReader.py

```diff
@@ -6,53 +6,383 @@
 
 import logging
 import os
+import time
+import copy
 from Utils.WebTools import getResponse
 from Utils.ConfigurationHandler import ConfigurationHandler
 
+from typing import List, Optional, Union, Any, Callable
+
 
 class ReqMgrReader(object):
     """
     _ReqMgrReader_
     General API for reading data from ReqMgr
     """
 
-    def __init__(self, url=None, logger=None, **contact):
-
+    def __init__(
+        self, url: str = None, logger: Optional[logging.Logger] = None, **contact
+    ) -> None:
+        # TODO: url is not being used
         try:
             configurationHandler = ConfigurationHandler()
-            self.reqmgrUrl = os.getenv('REQMGR_URL', configurationHandler.get('reqmgr_url'))
+            self.reqmgrUrl = os.getenv(
+                "REQMGR_URL", configurationHandler.get("reqmgr_url")
+            )
+            self.reqmgrEndpoint = {
+                "req": "/reqmgr2/data/request/",
+                "info": "/reqmgr2/data/info/",
+                "agentConfig": "/reqmgr2/data/wmagentconfig/",
+                "splitting": "/reqmgr2/data/splitting/",
+                "couchdb": "/couchdb/reqmgr_workload_cache/",
+            }
+
             logging.basicConfig(level=logging.INFO)
             self.logger = logger or logging.getLogger(self.__class__.__name__)
-        except Exception as e:
+
+        except Exception as error:
             msg = "Error initializing ReqMgrReader\n"
-            msg += "%s\n" % format(e)
+            msg += f"{error}\n"
             raise Exception(msg)
 
-    def getWorkflowByCampaign(self, campaign, details=False):
+    def _runWithRetries(
+        self,
+        fcn: Callable,
+        fcn_pargs: list,
+        fcn_args: dict = {},
+        default: Optional[Any] = None,
+        tries: int = 10,
+        wait: int = 5,
+    ) -> Any:
+        # TODO: is there a better place to put this function?
+        """
+        The function to run a given function with retries
+        :param fcn: function
+        :param fcn_pargs: function arguments
+        :param fnc_args: optional function arguments
+        :default: default value to return in case all the tries have failed, raise Exception o/w
+        :tries: number of tries
+        :wait: wait time between tries
+        :return: function output, deafault value o/w
+        """
+        for i in range(tries):
+            try:
+                return fcn(*fcn_pargs, **fcn_args)
+
+            except Exception as error:
+                self.logger.error(
+                    f"Failed to get to run function {fcn.__name__} with arguments {fcn_pargs} and {fcn_args} on try #{i+1} of {tries}"
+                )
+                self.logger.error(str(error))
+                time.sleep(wait)
+
+        if default:
+            return default
+        raise Exception("NoDefaultValue")
+
+    def getWorkflowSchema(
+        self,
+        wf: str,
+        drop_null: bool = False,
```

Comment:
I would remove this argument, I do not see a use case of it. 

Another general comment: 
1. We should avoid a function having too many arguments. It's not a good practice. If a function needs to take more than 2 arguments, we should ask ourselves:
    1. Are we sure that this function is atomic enough? A function should be responsible for just one thing. 
    2. If the answer is yes, keep all arguments under a dict (called **kwargs, or params)

Reply:
If we can achieve this, we can keep the function signature in one line. It would be much readable.